### PR TITLE
JS: Keep the position of end-of-line comments in binary expressions

### PR DESCRIPTION
**Pull request by shisama**
[Playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEA3AhgJwATuwXmwCNsAfU7AegCprsABGATwAc5tgBnGTASygDmAX2zVK2ADpRsM7GADcIADQgILGL2idkoLJggB3AApYE2lOlQReAE2UgimdGADWcGAGUWz-gOQ8AVzgVAAsYAFsAGwB1EN54Tm8wOA8zeN5UeKZkcE5tFX5OOEwYIycBcPRkADN0SKKVACtOAA8AISdXdw90cLgAGX44GrqGkGaWj19IuABFAIh4EfrgkG9MIswconQiOEj7Fj5YaNsYEOQADgAGFSOIIuinFhyjuE3UYZUARwX4MrU5hA6E4AFooHA4DYofZMHBfrw4WV0BUqkhaisVEVwrx-JggljpnM-sN0aNVjBdqcbOdkAAmFQ8dC8SK+ADCEHClRyUGgXxAASKABVduYMWNUEEAJJQaGwDxgPjqACCso8zBmyyKQiEQA)

input
```javascript
var a = b || /** @type {string} */
    (c);

var a = b + /** @type {string} */ 
    c;
```

stable
```javascript
var a = b /** @type {string} */ || c;

var a = b /** @type {string} */ + c;
```

expected
```javascript
var a = b || /** @type {string} */ (c);

var a = b + /** @type {string} */ (c);
```

Block comment between logical expression (`||`, `&&`) moves from left to right.
This causes error with google-closure-compiler.

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: changelog_unreleased/javascript/pr-7787.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#### JS: Keep comments position for LogicalExpression ([#7787](https://github.com/prettier/prettier/pull/7787) by [@shisama](https://github.com/shisama))
```
Comment (fisker) [inline changelog_unreleased/javascript/pr-7787.md]: This `JS:` should be a error caught by `lint:changelog`, can you update this code

https://github.com/prettier/prettier/blob/aacb5fa75f32c9d596a4e50d122914f14b87c9b2/scripts/lint-changelog.js#L113

Maybe like

```js
if ([...CHANGELOG_CATEGORIES, 'js'].includes(categoryInTitle.toLowerCase())) {
```

to forbid title startsWith `JS:`?

File: src/language-js/comments.js [hunk 2]
```diff
@@ -884,6 +892,30 @@ function handleTSMappedTypeComments(
...
+  comment,
+  options
+) {
+  if (enclosingNode && enclosingNode.type === "LogicalExpression") {
+    if (
+      followingNode &&
+      ((["babel", "babel-ts"].includes(options.parser) &&
+        comment.type === "CommentBlock") ||
+        (["flow", "typescript"].includes(options.parser) &&
+          comment.type === "Block")) &&
```
Comment (sosukesuzuki) [inline src/language-js/comments.js]: Can we use `isBlockComment` function instead of these logic?

File: src/language-js/comments.js [hunk 3]
```diff
@@ -884,6 +892,30 @@ function handleTSMappedTypeComments(
...
 
+function handleLogicalExpression(
+  text,
+  enclosingNode,
+  precedingNode,
+  followingNode,
+  comment,
+  options
+) {
+  if (enclosingNode && enclosingNode.type === "LogicalExpression") {
```
Comment (sosukesuzuki) [inline src/language-js/comments.js]: This bug happens for `BinaryExpression`, not only `LogicalExpression`.

[Playground Link](https://deploy-preview-7787--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEA3AhgJwATuwXmwCNsBqbAegCorsABGATwAc5tgBnGTASygHMAvtioVsAHSjZp2MAG4QAGhARmMHtA7JQWTBADuABSwItKdABt96RluVFM6MAGs4MAMrMnffsm4BXOGUACxgAWwsAdWCeeA4vMDh3U1ieVFjGZHAOOxA+DjhMGENHfjD0ZAAzSwLlACsOAA8AIUcXN3d0MLgAGT44KpqgkAbG9x8LOABFfwh4QYtakC9MAswsonQiOAslZd5YSJ4AExhg5AAOAAZlZj0CyMdmLLu4NdQB5QBHWfgS1TMIHQHAAtFA4HBjpC9pg4D8eLCSugyhUkNVFsMCmEeH5MIFlBwJtNfgM0UNlDAtkdTuckAAmCmOHgWHwAYQgYXKWTeAFY9v4CgAVLZmdFLVCBACSUChsHcYF4agAgjL3ExJgsCoJBEA)

**Input:**

```js
var a = b + /** @type {string} */ 
    c;
```

**Output:**

```js
var a = b /** @type {string} */ + c;
```

**Comment by thorn0** [comment]
```js
/** @type {string} */ c
```
vs
```js
/** @type {string} */ (c)
```
Do closure type casts work without parens?

**Comment by shisama** [comment]
@thorn0 It needs parens.

File: src/language-js/comments.js (hunk 3 above)
Comment (shisama) [reply src/language-js/comments.js]: Thanks. I will add `handleBinaryExpression` function.

**Comment by j-f1** [comment]
New output as of v2.0:

**Prettier 2.0.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEA3AhgJwATuwXmwCNsAfU7AegCprsABGATwAc5tgBnGTASygDmAX2zVK2ADpRsM7AAowASgDcUqRhx5CJclVoNmbDtz6CRYmQpUgANCAgsYvaJ2SgsmCAHcAClgSuKOioELwAJrYgRJjoYADWcDAAyiyx-ALIPACucHYAFjAAtgA2AOp5vPCcqWBwSQGVvKiVTMjgnK52-JxwmDA+MQKF6MgAZujFPXYAVpwAHgBCMfGJSeiFcAAy-HBjE1Mgs3NJ6cVwAIpZEPB7k7kgqZg9mG1E6ERwxZEspjCl4TA8sgABwABjsPwgPVKMRYbR+cGeqF2dgAjld4AMHIEQOhOABaKBwOBhEmRTBwdG8CkDdBDEZIcZ3Ow9Qq8TKYHIs04XDG7Rn7e4wd7-MKA5AAJjsPHQvGK6QAwhBCsM2lBoCiQFkegAVd6BJkHVA5ACSUFJsCSYD4jgAguakswzrcekIhEA)
```sh
--arrow-parens avoid
--parser babel
--trailing-comma none
```

**Input:**
```jsx
var a = b || /** @type {string} */ 
    (c);

var a = b || /** @type {string} */  (c);
```

**Output:**
```jsx
var a =
  b ||
  /** @type {string} */

  (c);

var a = b || /** @type {string} */ (c);

```

**Comment by thorn0** [comment]
@j-f1 Please notice that there is a space at the end of the first line of the input. That's what we get if we remove it:

**Prettier 2.0.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEA3AhgJwATuwXmwCNsAfU7AegCprsABGATwAc5tgBnGTASygDmAX2zVKAHSjZp2ABRgAlAG5JkjDjyES5KrQbM2Hbn0Eix0+cpAAaEBBYxe0TslBZMEAO4AFLAhco6KgQvAAmNiBEmOhgANZwMADKLDH8Asg8AK5wtgAWMAC2ADYA6rm88JwpYHCJ-hW8qBVMyOCcLrb8nHCYMN7RAgXoyABm6EXdtgBWnAAeAELRcQmJ6AVwADL8cKPjkyAzs4lpRXAAipkQ8LsTOSApmN2YrUToRHBFESwmMCVhMLlkAAOAAMtm+EG6JWiLFa3zgT1QO1sAEdLvB+vYAiB0JwALRQOBwULEiKYOBo3jk-roQbDJBjW62boFXgZTDZZknc7onYMvZ3GBvP6hAHIABMth46F4RTSAGEIAUhq0oNBkSBMt0ACpvAKM-aobIASSgJNgiTAfAcAEEzYlmKcbt0hEIgA)
```sh
--arrow-parens avoid
--parser babel
--trailing-comma none
```

**Input:**
```jsx
var a = b || /** @type {string} */
    (c);

var a = b || /** @type {string} */  (c);
```

**Output:**
```jsx
var a = b || /** @type {string} */ (c);

var a = b || /** @type {string} */ (c);

```

**Comment by shisama** [comment]
@sosukesuzuki @j-f1 @thorn0 
Thank you for your comments. As you pointed out, I think that the output([Playground](https://deploy-preview-7787--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEA3AhgJwATuwXmwCNsAfU7AegCprsABGATwAc5tgBnGTASygDmAX2zVK2ADpRsM7GADcIADQgILGL2idkoLJggB3AApYE2lOgA2B9E20qimdGADWcGAGUWz-gOQ8AVzgVAAsYAFtLAHUQ3nhObzA4DzM43lQ4pmRwTnsQfk44TBgjJwFw9GQAMytClQArTgAPACEnV3cPdHC4ABl+OGra4JBGpo9fSzgARQCIeCHLOpBvTELMbKJ0IjhLZRW+WCjeABMYEOQADgAGFRZ9QqinFmz7uHXUQZUARzn4UrU5hA6E4AFooHA4CcoftMHBfrw4aV0OVKkgaksRoVwrx-JggipOJMZn9BujhioYNtjmcLkgAEyUpy8Sy+ADCEHCFWy7wArPsAoUACrbcwY5aoIIASSg0NgHjAfHUAEFZR5mFNFoUhEIgA)) is ugly when there are spaces at the end of the line.

When I changed the below code, the case is fixed.

https://github.com/prettier/prettier/blob/763a1e2f2455c594d4b3db092b90711ce44f429b/src/main/comments.js#L524-L526

```diff
-      if (index !== false && hasNewline(text, index)) {
+      if (
+        index !== false &&
+        hasNewline(text, index) &&
+        !hasSpaces(text, index)
+      ) {
         leadingParts.push(hardline);
       }
```

However, two tests failed. Do you have any ideas?

<details>

```

Summary of all failing tests
 FAIL  tests/assignment_comments/jsfmt.spec.js
  ● function.js

    expect(received).toMatchSnapshot()

    Snapshot name: `function.js 1`

    - Snapshot  - 6
    + Received  + 0

    @@ -81,18 +81,15 @@

      f4 = () => {}; // Comment

      f5 =
        // Comment
    -
        () => {};

      f6 =
        /* comment */
    -
        // Comment
    -
        () => {};

      let f1 = (
        //comment
        a = b
    @@ -108,16 +105,13 @@

      let f4 = () => {}; // Comment

      let f5 =
        // Comment
    -
        () => {};

      let f6 =
        /* comment */
    -
        // Comment
    -
        () => {};

      ================================================================================

       98 |           )
       99 |         )
    > 100 |       ).toMatchSnapshot();
          |         ^
      101 |     });
      102 | 
      103 |     const parsersToVerify = parsers.slice(1);

      at Object.test (tests_config/run_spec.js:100:9)

 FAIL  tests/insert-pragma/js/jsfmt.spec.js
  ● module-with-late-docblock.js

    expect(received).toMatchSnapshot()

    Snapshot name: `module-with-late-docblock.js 1`

    - Snapshot  - 1
    + Received  + 0

    @@ -21,9 +21,8 @@
      }

      /**
       * Some notes that should not be appended to
       */
    -
      const fruit = "tomatoes";

      ================================================================================
```

</details>

**Comment by sosukesuzuki** [comment]
@shisama Sorry for delay. I think we can ignore those changes and update snapshots them.
/cc @thorn0 @j-f1 How do you think?

**Comment by shisama** [comment]
@sosukesuzuki Thank you.
@thorn0 @j-f1 Will you please tell me how you think?

File: src/language-js/comments.js [hunk 4]
```diff
@@ -884,6 +900,48 @@ function handleTSMappedTypeComments(
...
+    ) {
+      addLeadingComment(followingNode, comment);
+      return true;
+    }
+  }
+  return false;
+}
+
+function handleBinaryExpression(
+  text,
```
Comment (thorn0) [inline src/language-js/comments.js:925]: `handleLogicalExpression` and `handleBinaryExpression` are almost identical. Please make them one function.

**Comment by thorn0** [comment]
The bug with trailing whitespace after end-of-line comments is a separate issue, not specific binary expressions. No need to fix it here.

**Comment by shisama** [comment]
@thorn0 Thank you for your comments. I will create another PR for #8068 and rebase this.

**Comment by shisama** [comment]
I rebased this and resolved the problem of the extra spaces after comment.

**Comment by shisama** [comment]
sencond format tests failed. So, tests for this are added to unstable tests (ref: https://github.com/prettier/prettier/pull/7565/ ).

```js
// original
a = b || // TODO this is a very very very very long comment that makes it go > 80 columns
c;

// first format
a =
  b || c; // TODO this is a very very very very long comment that makes it go > 80 columns

// second format
a = b || c; // TODO this is a very very very very long comment that makes it go > 80 columns
```

File: src/language-js/comments.js (hunk 4 above)
Comment (shisama) [reply src/language-js/comments.js:922]: @thorn0 OK, I fixed it on https://github.com/prettier/prettier/pull/7787/commits/51f1b50df4ce5b5f2da8211aad77c3d210628fe8
Though I named the function `handleEndOfLineBlockComments`, I am worrying if it is suitable for it. Would you give me some advice if you have any idea?

Comment (thorn0) [reply src/language-js/comments.js:922]: The binary expression name was good. Logical expressions are binary too.

File: src/language-js/comments.js [hunk 5]
```diff
@@ -911,6 +918,27 @@ function handleTSMappedTypeComments(
...
+  comment,
+  options
+) {
+  if (
+    enclosingNode &&
+    (enclosingNode.type === "LogicalExpression" ||
+      enclosingNode.type === "BinaryExpression") &&
+    followingNode &&
+    isBlockComment(comment) &&
+    privateUtil.hasNewline(text, options.locEnd(comment))
```
Comment (thorn0) [inline src/language-js/comments.js:934]: Why is this condition needed? This function is called from `handleEndOfLineComment`. Isn't the comment by this point already classified as end-of-line?

**Comment by thorn0** [comment]
Please move the tests for single-line comments to a separate file and rename the files to make it clear that they have to do with binary expressions.

File: src/language-js/comments.js (hunk 4 above)
Comment (shisama) [reply src/language-js/comments.js:916]: Thanks. I fixed on https://github.com/prettier/prettier/pull/7787/commits/3f743fd26274e3cddd97aa223c5e5524d5dddd38.

File: src/language-js/comments.js (hunk 5 above)
Comment (shisama) [reply src/language-js/comments.js]: That's exactly. This is fixed on https://github.com/prettier/prettier/pull/7787/commits/a6dabaf7d0a05b023225be76307104f2a361d282.

**Comment by shisama** [comment]
@thorn0 Thanks for your review. I fixed them. Please take a look again.

File: tests_config/run_spec.js [hunk 6]
```diff
@@ -25,6 +25,7 @@ const unstableTests = new Map(
     "class_comment/comments.js",
     ["comments/dangling_array.js", (options) => options.semi === false],
     ["comments/jsx.js", (options) => options.semi === false],
+    "comments/binary-expressions-single-comments.js",
```
Comment (fisker) [inline tests_config/run_spec.js:28]: I hate this list growing, but not related to this PR, good job.

Comment (shisama) [reply tests_config/run_spec.js:28]: I agree. I have ideas.

1. Unstable test files have `// Flag: unstable test` comment on the top.
1. Make `tests/unstable` directory and move unstable test files into it.
1. Move this list to another file. (e.g. `unstable_tests.json`) This is easiest to implement.

Do you have any ideas?

Comment (fisker) [reply tests_config/run_spec.js:28]: It's okay to keep this list here, easy to trace.

we need kill them all, and we will.

**Comment by thorn0** [comment]
Aren't there open issues this PR fixes?

**Comment by shisama** [comment]
I don't know any issue related to this PR. I think there aren't.

**Pull request merged**